Work in progress on an AutoEncoder example

tennlee opened the pull request:
Model is functional but performance needs optimisation. (e.g. add skip connections, improve the architecture etc). This will be done in a subsequent PR.

tennlee committed:
Work in progress on an AutoEncoder example
Model is functional but performance not very impressive

stevehadd commented:
Things to work:
* add more plotting of prediction and target side-by-side
* try things to improve thing
  * additional epochs
  * additional layers
  * larger latent space
* plot some accuracy scores for images using different metrics
  * Kernel Inception Distance
  * Structural Similarity Index 
    * https://scikit-image.org/docs/0.25.x/auto_examples/transform/plot_ssim.html 
    * https://stackoverflow.com/questions/53956932/use-pytorch-ssim-loss-function-in-my-model 
  * Learned Perceptual Image Patch Similarity (LPIPS)
  * spectral loss? 
* Try structural similarity index based loss function (1 - the SSIM)

Useful links:
* some articles on CNNs with satellite imagery
  * https://ieeexplore.ieee.org/document/9883256 
  * https://arxiv.org/html/2505.00040v1 
* metrics
  * https://huggingface.co/blog/PrunaAI/objective-metrics-for-image-generation-assessment 
  * https://arxiv.org/html/2505.07175v1

tennlee committed:
Update pipeline and controller classes to allow for skipping missing data through allowed exceptions in the iterator
Expand the exception handler to permit an unlimited exception count for allowed exceptions

tennlee committed:
Code autoformatting

tennlee committed:
Add imported class to __all__ for module

tennlee committed:
Additional reformatting following ruff fix
